stake-pool/program/src/processor.rs

```diff
@@ -560,79 +561,198 @@ impl PrintProgramError for Error {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::instruction::{initialize, Fee, InitArgs};
-    use core::mem::size_of;
+    use crate::instruction::{deposit, initialize, Fee, InitArgs};
     use solana_program::{
         account::Account, account_info::create_is_signer_account_infos, instruction::Instruction,
-        program_pack::Pack, rent::Rent, sysvar::rent,
+        native_token::sol_to_lamports, program_pack::Pack, program_stubs, rent::Rent, sysvar::rent,
+    };
+    use spl_token::{
+        instruction::{initialize_account, initialize_mint},
+        processor::Processor as TokenProcessor,
+        state::{Account as SplAccount, Mint as SplMint},
     };
-    use spl_token::instruction::{initialize_account, initialize_mint};
 
-    // Test program id for the stake-pool program.
+    /// Test program id for the stake-pool program.
     const STAKE_POOL_PROGRAM_ID: Pubkey = Pubkey::new_from_array([2u8; 32]);
 
-    // Test program id for the token program.
+    /// Test program id for the token program.
     const TOKEN_PROGRAM_ID: Pubkey = Pubkey::new_from_array([1u8; 32]);
 
+    /// Actual stake account program id, used for tests
+    fn stake_program_id() -> Pubkey {
+        "Stake11111111111111111111111111111111111111"
+            .parse::<Pubkey>()
+            .unwrap()
+    }
+
+    struct TestSyscallStubs {}
+    impl program_stubs::SyscallStubs for TestSyscallStubs {
+        fn sol_invoke_signed(
+            &self,
+            instruction: &Instruction,
+            account_infos: &[AccountInfo],
+            signers_seeds: &[&[&[u8]]],
+        ) -> ProgramResult {
+            info!("TestSyscallStubs::sol_invoke_signed()");
+
+            let mut new_account_infos = vec![];
+            for meta in instruction.accounts.iter() {
+                for account_info in account_infos.iter() {
+                    if meta.pubkey == *account_info.key {
+                        let mut new_account_info = account_info.clone();
+                        for seeds in signers_seeds.iter() {
+                            let signer =
+                                Pubkey::create_program_address(seeds, &STAKE_POOL_PROGRAM_ID)
+                                    .unwrap();
+                            if *account_info.key == signer {
+                                new_account_info.is_signer = true;
+                            }
+                        }
+                        new_account_infos.push(new_account_info);
+                    }
+                }
+            }
+
+            match instruction.program_id {
+                TOKEN_PROGRAM_ID => invoke_token(&new_account_infos, &instruction.data),
+                pubkey => {
```

Comment:
Could you have this match arm work like the token one? then you could have something like:
```rust
let stake_id = stake_program_id();
match instruction.program_id {
    TOKEN_PROGRAM_ID => ...,
    stake_id => ...,
    _ => Err
}
```

Reply:
Stake id is not a constant, so I could not really do this in the same way as with token program id, may be it will be better after direct access to stake program is added for testing (the one promised by mvines)